app/controllers/incomeSources/add/AddIncomeSourceStartDateCheckController.scala

```diff
@@ -177,39 +175,63 @@ class AddIncomeSourceStartDateCheckController @Inject()(authenticate: Authentica
                               successUrl: String,
                               incomeSourceStartDate: String,
                               incomeSourceType: IncomeSourceType)
-                             (implicit mtdItUser: MtdItUser[_]): Result = {
+                             (implicit mtdItUser: MtdItUser[_]): Future[Result] = {
 
     val formResponse: Option[String] = validForm.toFormMap(form.response).headOption
 
     (formResponse, incomeSourceType) match {
       case (Some(form.responseNo), _) =>
-        Redirect(backUrl)
-          .removingFromSession(
-            incomeSourceType.startDateSessionKey
-          )
+        removeSessionData(Seq(incomeSourceType.startDateSessionKey), Redirect(backUrl), showInternalServerError(isAgent))
       case (Some(form.responseYes), SelfEmployment) =>
-        Redirect(successUrl)
-          .addingToSession(
-            SessionKeys.addBusinessAccountingPeriodStartDate -> incomeSourceStartDate,
-            SessionKeys.addBusinessAccountingPeriodEndDate -> dateService.getAccountingPeriodEndDate(incomeSourceStartDate)
-          )
+        setSessionData(Seq(
+          (addBusinessAccountingPeriodStartDate, incomeSourceStartDate),
+          (addBusinessAccountingPeriodEndDate, dateService.getAccountingPeriodEndDate(incomeSourceStartDate))
+        )
+          , Redirect(successUrl)
+        ) map {
+          case Left(_) => showInternalServerError(isAgent)
+          case Right(result) => result
+        }
       case (Some(form.responseYes), _) =>
-        Redirect(successUrl)
+        Future.successful(Redirect(successUrl))
       case _ =>
         Logger("application").error(s"[AddIncomeSourceStartDateCheckController][handleValidForm] - Unexpected response, isAgent = $isAgent")
-        showInternalServerError(isAgent)
+        Future.successful(showInternalServerError(isAgent))
     }
   }
 
-  private def getAndValidateStartDate(incomeSourceType: IncomeSourceType)
-                                     (implicit user: MtdItUser[_]): Either[Throwable, String] = {
+  private def removeSessionData(keys: Seq[String], successResult: Result, errorResult: Result)(implicit mtdItUser: MtdItUser[_]): Future[Result] = {
+    sessionService.remove(keys, successResult) map {
+      case Left(_) => errorResult
+      case Right(result) => result
+    }
+  }
+
+  private def setSessionData(keyValuePairs: Seq[(String, String)], result: Result)(implicit mtdItUser: MtdItUser[_]): Future[Either[Throwable, Result]] = {
```

Comment:
you might consider using foldLeft instead of this custom list processing which is really redundant => you can use standard scala collection methods => ie

```
 keyValuePairs.foldLeft[Future[Either[Throwable, Result]]](Future{Right(redirect)}) {(acc, keyValue) =>
          val result = for {
            a <- acc
          } yield a match {
            case Right(res) =>
              sessionService.set(keyValue._1, keyValue._2, res)
            case Left(ex) =>
              Future{ Left(ex) }
          }
          result.flatten
        }
```

Reply:
https://github.com/hmrc/income-tax-view-change-frontend/pull/1891/files